man/fpicprog.1.txt
:version: <VERSION>
:summary: PIC programmer software using FTDI chips

NAME
====

fpicprog - PIC programmer software using FTDI chips

SYNOPSIS
========

fpicprog --action=<action> [_OPTIONS_]

DESCRIPTION
===========

fpicprog is a software driver for FTDI USB chips like the FT232RL to program
Microchip PIC\(rg chips.

OPTIONS
=======

fpicprog accepts the following options:

*--action*=_action_::
	Instruct fpicprog what to do. Available actions are _erase_,
	_dump-program_, _write-program_, _identify_, and _list-programmers_.
	When using _erase_, _dump-program_ or _write-program_, the --sections flag
	can be used to indicate which sections to operate on. For _write-program_
	and _dump-program_ an empty flag means all sections, while for erase an
	explicit --sections=all must be passed. The _list-programmers_ action
	identifies the available programmers connected to the computer. (See
	PROGRAMMERS below for information on how to make the programmers available
	to fpicprog). The _identify_ action attempt to identify the PIC chip
	connected to the programmer. This does not work for all devices, as not
	all devices provide a device ID. Also note that you have to select the
	appropriate device family (see --family) for this to work.
*--help*::
	Display a comprehensive help message. This also lists several options
	which have been omitted here for brevity.
*--section*=_sections_::
	Selects the sections to operate on. Valid names are either _all_ or a comma
	separated list of one or more of _flash_, _user-id_, _config_, _eeprom_.
	Note that not all devices/families support operating on a subset of the
	available sections.
*--family*=_family_::
	Selects the device family the connected device belongs to. This should be
	one of pic10, pic12, pic16, pic16-new or pic18. Other values are accepted,
	but are either for incomplete implementations, or for high-voltage
	programming, which is not officially supported by fpicprog at this time.
*--device*=_device_::
	Selects the device to program. Acceptable values are determined by the
	family. Typically, this is the device name without the leading PIC, and
	using capital letters. E.g. 18F45K50. See the device list files for
	possible values.
*--device_db*=_device database path_::
	Device DB file to load. The default value depends on the --family flag,
	and can be seen by looking at the --help output.
*--input*=_file name_::
	Hex file to read for writing to the device.
*--output*=_file name_::
	Location to write the Hex file to when dumping a program.
*--erase_mode*=_mode_::
	Mode of erasing the device to use. Either _chip_, _section_ or _none_. The
	default mode is _chip_, meaning that the entire chip will be erased when
	selection the actions _erase_ or _write-program_.

*--ftdi_product_id*=_pid_::
	Select the USB Product ID of the programmer to open. By default fpicprog
	will look for a programmer with USB Product ID 0x6001.
TODO: list the other ftdi_ options, especially those changing the pins.

PROGRAMMERS
===========

fpicprog was designed to work with FTDI FT232RL chips. However, several other
FTDI chips provide the required functionality (i.e. synchronous bit-bang mode)
as well. This includes the FT232RQ, FT245RL, FT245RQ, FT4232H and the FT2232
series of devices. These use different USB Product IDs (0x6010, 0x6011, 0x6014
and 0x6015).

On Linux systems, by default the programmers typically are only accessible by
root. To make fpicprog able to access the programmers as a regular user, use
udev rules to set the permissions on the programmer. The following example
makes the programmers (with USB Product ID 0x6001) accessible to all users:

SUBSYSTEM=="usb", ATTR{idVendor}=="0403", ATTR{idProduct}=="6001", MODE="0666"

Depending on the device you are using, you may need to set the product ID
to 6001, 6010, 6011, 6014 or 6015

BUGS
====

If you think you have found a bug, please check that you are using the latest
version of fpicprog <http://os.ghalkes.nl/fpicprog.html>. A bugtracker is
available at <https://github.com/gphalkes/fpicprog/issues>. When reporting bugs,
please include a minimal example that demonstrates the problem.

AUTHOR
======

G.P. Halkes <tilde@ghalkes.nl>

COPYRIGHT
=========

Copyright \(co 2018 G.P. Halkes +
fpicprog is licensed under the GNU General Public License version 3. +
For more details on the license, see the file COPYING in the documentation
directory. On Un*x systems this is usually /usr/share/doc/fpicprog.


src/fpicprog.cc
/* Copyright (C) 2016 G.P. Halkes
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 3, as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <algorithm>
#include <cerrno>
#include <cstring>
#include <gflags/gflags.h>
#include <set>
#include <vector>

#include "controller.h"
#include "driver.h"
#include "high_level_controller.h"
#include "pic16controller.h"
#include "pic18controller.h"
#include "pic24controller.h"
#include "picnew8bitcontroller.h"
#include "program.h"
#include "sequence_generator.h"
#include "status.h"
#include "strings.h"

DEFINE_string(
    action, "",
    "Action to perform. One of erase, dump-program, write-program, identify, list-programmers. "
    "When using erase, dump-program or write-program, the --sections flag can be used to "
    "indicate which sections to operate on. For write-program and dump-program an empty "
    "flag means all sections, while for erase an explicit --sections=all must be passed.");
DEFINE_string(sections, "",
              "Comma separate list of sections to operate on. Possible values: either all "
              "or a combination of flash, user-id, config, eeprom.");
DEFINE_string(family, "",
              "Device family to use. One of pic10, pic10-baseline, pic12, pic12-baseline, pic16, "
              "pic16-baseline, pic16-new, pic18, pic18-new.");
DEFINE_string(device, "",
              "Exact device name. Required for devices which don't provide a device ID. Devices "
              "which have a device ID should be detectable using the identify action.");

DEFINE_string(output, "", "File to write the Intel HEX data to (--action=dump-program).");
DEFINE_string(input, "", "Intel HEX file to read and program. (--action=write-program)");
DEFINE_string(erase_mode, "chip", "Erase mode for writing. One of chip, section, none.");
DEFINE_string(device_db, "",
              "Device DB file to load. Defaults to "
#if defined(DEVICE_DB_PATH)
              DEVICE_DB_PATH "/<family>.lst.");
#else
              "<path to binary>/device_db/<family>.lst.");
#endif

static EraseMode ParseEraseMode() {
  if (FLAGS_erase_mode == "chip") {
    return CHIP_ERASE;
  } else if (FLAGS_erase_mode == "section") {
    return SECTION_ERASE;
  } else if (FLAGS_erase_mode == "none") {
    return NO_ERASE;
  } else {
    fatal("No such erase mode '%s'\n", FLAGS_erase_mode.c_str());
  }
}

int main(int argc, char **argv) {
  google::ParseCommandLineFlags(&argc, &argv, true);

  std::unique_ptr<Driver> driver = Driver::CreateFromFlags();
  if (FLAGS_action.empty()) {
    fatal("No action specified\n");
  } else if (FLAGS_action == "list-programmers") {
    std::vector<std::string> devices;
    CHECK_OK(driver->List(&devices));
    for (const auto &device : devices) {
      printf("Device:\n%s", device.c_str());
    }
    exit(0);
  }

  // FIXME: move initialization of the device DB to a separate function to be called from both here
  // and testgen.
  std::unique_ptr<Controller> controller;
  std::unique_ptr<DeviceDb> device_db;
  if (FLAGS_family.empty()) {
    fatal("--family must be specified\n");
  } else if (FLAGS_family == "pic18") {
    std::unique_ptr<Pic18SequenceGenerator> sequence_generator(new Pic18SequenceGenerator);
    controller.reset(new Pic18Controller(std::move(driver), std::move(sequence_generator)));
    device_db = std::make_unique<DeviceDb>(1, 1, Datastring{0xff},
                                           [](const Datastring16 &) { return Status::OK; });
  } else if (FLAGS_family == "pic18-new") {
    std::unique_ptr<PicNew8BitSequenceGenerator> sequence_generator(
        new PicNew8BitSequenceGenerator);
    controller.reset(new PicNew8BitController(std::move(driver), std::move(sequence_generator),
                                              PicNew8BitController::PIC18NEW));
    device_db = std::make_unique<DeviceDb>(1, 1, Datastring{0xff},
                                           [](const Datastring16 &) { return Status::OK; });
  } else if (FLAGS_family == "pic10" || FLAGS_family == "pic12" || FLAGS_family == "pic16") {
    std::unique_ptr<Pic16SequenceGenerator> sequence_generator(new Pic16SequenceGenerator);
    controller.reset(new Pic16MidrangeController(std::move(driver), std::move(sequence_generator)));
    device_db =
        std::make_unique<DeviceDb>(2, 2, Datastring{0xff, 0x3f}, [](const Datastring16 &sequence) {
          return Pic16SequenceGenerator::ValidateSequence(sequence);
        });
  } else if (FLAGS_family == "pic10-small" || FLAGS_family == "pic12-small" ||
             FLAGS_family == "pic16-small") {
    std::unique_ptr<Pic16SequenceGenerator> sequence_generator(new Pic16SequenceGenerator);
    controller.reset(new Pic16BaselineController(std::move(driver), std::move(sequence_generator)));
    device_db =
        std::make_unique<DeviceDb>(2, 2, Datastring{0xff, 0x0f}, [](const Datastring16 &sequence) {
          return Pic16SequenceGenerator::ValidateSequence(sequence);
        });
  } else if (FLAGS_family == "pic16-new") {
    std::unique_ptr<PicNew8BitSequenceGenerator> sequence_generator(
        new PicNew8BitSequenceGenerator);
    controller.reset(new PicNew8BitController(std::move(driver), std::move(sequence_generator),
                                              PicNew8BitController::PIC16NEW));
    device_db =
        std::make_unique<DeviceDb>(2, 2, Datastring{0xff, 0x3f}, [](const Datastring16 &sequence) {
          return Pic16SequenceGenerator::ValidateSequence(sequence);
        });
  } else if (FLAGS_family == "pic24") {
    std::unique_ptr<Pic24SequenceGenerator> sequence_generator(new Pic24SequenceGenerator);
    controller.reset(new Pic24Controller(std::move(driver), std::move(sequence_generator)));
    device_db = std::make_unique<DeviceDb>(4, 2, Datastring{0xff, 0xff, 0xff, 0x00},
                                           [](const Datastring16 &) { return Status::OK; });
  } else {
    fatal("Unknown device family %s.\n", FLAGS_family.c_str());
  }

  std::string filename = FLAGS_device_db;
  if (filename.empty()) {
#if defined(DEVICE_DB_PATH)
    filename = strings::Cat(DEVICE_DB_PATH, "/", FLAGS_family, ".lst");
#else
    filename = strings::Cat(Dirname(argv[0]), "/device_db/", FLAGS_family, ".lst");
#endif
  }
  CHECK_OK(device_db->Load(filename));

  HighLevelController high_level_controller(std::move(controller), std::move(device_db));
  if (!FLAGS_device.empty()) {
    high_level_controller.SetDevice(FLAGS_device);
  }

  if (FLAGS_action == "erase") {
    if (FLAGS_sections.empty()) {
      fatal("Erase requires setting --sections\n");
    } else if (FLAGS_sections == "all") {
      CHECK_OK(high_level_controller.ChipErase());
    } else {
      CHECK_OK(high_level_controller.SectionErase(ParseSections(FLAGS_sections)));
    }
  } else if (FLAGS_action == "dump-program") {
    Program program;
    CHECK_OK(high_level_controller.ReadProgram(ParseSections(FLAGS_sections), &program));
    FILE *out = fopen(FLAGS_output.c_str(), "w+b");
    if (!out) {
      fatal("Could not open file '%s': %s\n", FLAGS_output.c_str(), strerror(errno));
    }
    WriteIhex(program, out);
    fclose(out);
  } else if (FLAGS_action == "write-program") {
    if (FLAGS_input.empty()) {
      fatal("--input is required for action write-program\n");
    }
    Program program;
    FILE *in = fopen(FLAGS_input.c_str(), "rb");
    if (!in) {
      fatal("Could not open file '%s': %s\n", FLAGS_input.c_str(), strerror(errno));
    }
    CHECK_OK(ReadIhex(&program, in));
    CHECK_OK(high_level_controller.WriteProgram(ParseSections(FLAGS_sections), program,
                                                ParseEraseMode()));
  } else if (FLAGS_action == "identify") {
    CHECK_OK(high_level_controller.Identify());
  } else {
    fatal("Unknown action '%s'\n", FLAGS_action.c_str());
  }
  return EXIT_SUCCESS;
}
